### ISSUE TITLE
Accessibility regression related to element roles

### ISSUE THREAD

**Issue by neefrehman** [recharts/recharts#3449]
- [x] I have searched the [issues](https://github.com/recharts/recharts/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
[https://master--63da8268a0da9970db6992aa.chromatic.com/?path=/story/api-chart-barchart--simple](https://master--63da8268a0da9970db6992aa.chromatic.com/?path=/story/api-chart-barchart--simple)

### Steps to reproduce
Create a chart that includes a `Bar`, `Line`, `Scatter` or `Sector`.

### What is expected?
The chart elements should have correct aria roles.

### What is actually happening?
Individual chart elements have an `"img"` role, which based on the WCAG spec should only be used for entire graphics that are single, indivisible elements.

| Environment | Info |
|---|---|
| Recharts | v2.4.3 |
| React | all |
| System | all |
| Browser | all |

---
We recently updated the version of Recharts we were using to fix a security vulnerability, and have found that our test suite (which uses axe for some automated accessibility testing) has started failing, with errors like the below:

```shell
Expected the HTML found at $('.recharts-bar.recharts-layer:nth-child(6) > .recharts-bar-rectangles.recharts-layer > .recharts-bar-rectangle.recharts-layer[role="img"]:nth-child(1)') to have no violations:

<g class="recharts-layer recharts-bar-rectangle" role="img"><path name="Previous year" fill="#BABBBD" width="19" height="348.5" x="139" y="21.500000000000007" radius="0" class="recharts-rectangle" d="M 139,21.500000000000007 h 19 v 348.5 h -19 Z"></path></g>

Received:

"<svg> elements with an img role must have an alternative text (svg-img-alt)"
Fix any of the following:
  Element has no child that is a title
  aria-label attribute does not exist or is empty
  aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
  Element has no title attribute

You can find more information on this issue here:
https://dequeuniversity.com/rules/axe/4.4/svg-img-alt?application=axeAPI
```

After some looking I can see that `"img"` roles were added in #2960, with good intentions, but from some research into chart accessibility I gather that this is the wrong role to use for individual svg elements in charts, or at least is used incorrectly as it is currently implemented. One reason, as the test failure above notes, is that the `"img"` role suggests to a browser that there should be some alt text.

A fix for this (if you want to keep the `"img"` role) would be to add `alt=""` to these elements so that assistive technology can ignore them, as illustrated by [point H67 of the WCAG spec](https://www.w3.org/TR/WCAG20-TECHS/H67.html).

A more appropriate fix, from what I understand, would be to use the ARIA `"graphics-symbol"` role, which is for atomic units within a `"graphics-document"` (the implicit role for an svg element). As well as this, elements that are more group-like, such as axes, could be given a `"graphics-object"` role. This would be a more involved change, and I'm certainly open to other ideas, as I may have misunderstood some of the information I've been reading.

I should also note that this change is only related to the role of the elements, which on it's own doesn't really improve the accessibility of the charts, as correct titles, labels and relationships need to be added. I understand there has already been some conversation on these points, so in the interest of a narrower discussion for this issue, I've focused only on roles.

Sources:

1. https://www.w3.org/TR/graphics-aria-1.0/#graphics-object
2. https://www.w3.org/TR/graphics-aria-1.0/#graphics-symbol
3. https://blog.tenon.io/accessible-charts-with-aria/

**Comment by ckifer** [comment]
Hey @neefrehman 

Thanks for the deep dive. We can add `alt=""` to the img roles for now. 

The graphics module for aria must be relatively new (looks like 2018 which is quite new) and there isn't much documentation on them besides the spec. Can we prove that these are supported by browsers? My IDE for example doesn't know that these new roles exist - they are also not on MDN.

Thanks! Feel free to open a PR to add the alt text

**Comment by neefrehman** [comment]
Thanks @ckifer. Yes it's quite new, and I've been struggling to understand browser support as well. There also seems to be some shared responsibility for graphics a11y between both this Graphics Module spec and the [SVG 2.0 spec](https://www.w3.org/TR/SVG2/), neither of which I can find any solid browser support info for! It's confusing. Though, as I read in the third source I shared, the [Graphics Accessibility API mappings](https://www.w3.org/TR/graphics-aam-1.0/#mapping_role) companion spec has [a test suite for browser compatibility](https://w3c.github.io/test-results/graphics-aam/). In this you can see [results](https://w3c.github.io/test-results/graphics-aam/all.html) from the time of publication (2018), showing that the roles are supported in all browsers with the exception of `graphics-document` and `graphics-object` in Chrome.

The proof is in the pudding, though, and I've just tried updating the DOM via DevTools (FF, Chrome & Safari) in the storybook page I linked. You can see below that all roles _are_ supported, including the ones that were failing in the linked test suite results. Note that unsupported roles fall back to the elements implicit role, or are marked as "Ignored". Feel free to try it yourself and compare the screen reader experience.

<img width="1321" alt="image" src="https://user-images.githubusercontent.com/29544565/224551599-f184f912-5c6a-46eb-9fc9-1f68366681df.png">

With this in mind, I think that leveraging these roles would be the best way forward for improving Recharts' accessibility. If supporting older browser is a concern, we could make this change a progressive enhancement by providing a fallback role. i.e. `role="graphics-symbol img"`.

Though with all the above being said, I've also just found a different ACT rule that would throw a spanner in the works for adopting these roles in isolation: [SVG element with explicit role has non-empty accessible name](https://www.w3.org/WAI/standards-guidelines/act/rules/7d6734/). It seems that—just like the ACT failure that caused me to post this issue—any adoption of explicit roles for elements must also be accompanied by accessible names via a `title`, `label`, or similar. This would increase the time required to implement, but I think it would doable to derive such a string from the input data. I would volunteer to look into this, if you agree it is the right approach.

Circling back to the `alt=""` fix, after more thought I don't think this would be right. [The HTML spec](https://html.spec.whatwg.org/multipage/embedded-content.html#the-img-element) says that this is appropriate for the following scenario:

>The image is either decorative or supplemental to the rest of the content, redundant with some other information in the document.

The element would be removed from the accessibility tree, similarly to `role="presentation"`. I've just tried doing this via DevTools and then going through the chart with a screen-reader. Some browsers read out "image" for each bar, and others don't read out anything at all. I would still consider this a regression compared to the elements not having an explicit role at all (in which case one "image" is read out for the whole chart), even though it would allow the ACT assertion to pass.

My vote would be to revert the introduction of `role="img"` and wait until we can confidently add accessible names to the elements, before reintroducing explicit roles.

**Comment by ckifer** [comment]
@neefrehman thanks for the deep dive. This info helps a ton!

I'm aligned for the most part with what you say here. Let's do some more testing, document those tests in this issue, then create issues that represent the work in smaller chunks. 

_Technically_ removing the img roles is a breaking change. Especially if consumers have tests that rely on it. So idk what to do there - we are about to start a 3.0 branch I think. 

If you want to volunteer to help with this work I think we can fit it in for sure. We just need to make sure that this time around we're actually testing using screen readers to make sure the results are what we intend.

**Comment by james2hey** [comment]
I've been running into a similar issue with accessibility tests failing for some recharts components as well. I installed the [accessibility testing](https://storybook.js.org/docs/react/writing-tests/accessibility-testing) storybook plugin in my branch to see what test failures we could surface, and have the following we could create some issues for to further investigate:

## Test failures:

### Landmarks should have a unique role or role/label/title (i.e. accessible name) combination

More details about the issue can be found here - [More info...](https://dequeuniversity.com/rules/axe/4.6/landmark-unique?application=axeAPI)
![Screenshot 2023-03-19 at 16 26 31](https://user-images.githubusercontent.com/41892498/226191149-44fecd88-eaae-42c7-8808-6b0f924463a7.png)

### Bad colour contrasts for default axis ticks:
More details about the issue can be found here - https://dequeuniversity.com/rules/axe/4.6/color-contrast?application=axeAPI

![Screenshot 2023-03-19 at 15 54 26](https://user-images.githubusercontent.com/41892498/226190559-73c8357d-f0cb-441f-b17a-49ecb4c0ff48.png)

### Svg Elements with an img role must have an alternative text
As documented in this issue.

It'd be great if we could create a ticket for each of these issues - let me know if we need any more info.

I also think it'd be nice to add this accessibility addon to storybook to prevent any regressions etc. Would you like me to open a separate PR/issue for this @ckifer? ([Changes for reference](https://github.com/recharts/recharts/compare/master...james2hey:recharts:storybook(a11y-testing))) The only reason I haven't done so already is that it introduces failing tests into the storybook test runner.

**Comment by ckifer** [comment]
@james2hey hey and thanks for this. Yeah I think both are great steps to take. Let's add the add-on in a different issue/PR as well as create issues for each that need explicitly addressed

**Comment by james2hey** [comment]
Yeah the add-on supports disabling on a story level, so we could do that case by case. And we could just make it an A.C. to remove the disabled flag for the story once the related accessibility issue is resolved

**Comment by rhuangabrielsantos** [comment]
Any update on the accessibility problem? @ckifer @neefrehman @james2hey

**Comment by ckifer** [comment]
@rhuangabrielsantos Not explicitly, accessibility will be an ongoing theme of improvements where possible (unbounded by time though since we only contribute as we can). If you have something specific you'd like to contribute towards improving feel free

**Comment by rhuangabrielsantos** [comment]
@ckifer https://github.com/recharts/recharts/pull/3614 I removed the role attribute until accessibility is worked out for all components, does that make sense?

**Comment by ckifer** [comment]
I've been trying to do a release roughly monthly. Probably next week.

**Comment by ckifer** [comment]
Removal of the role released in 2.7, leaving this open because it has good info in it. @neefrehman do you want to rename this?

**Comment by gcallaghan-nd** [comment]
I'm seeing the same issue in this file: https://github.com/recharts/recharts/blob/bb60505f46ff932851ce67f14299f1209d8a42df/src/cartesian/Line.tsx#L341

I'm getting the same ADA warning. Would it make sense to remove the role here as well?

**Comment by ckifer** [comment]
Yeah probably - @gcallaghan-nd want to raise a PR?

**Comment by gcallaghan-nd** [comment]
Sure, I'll get one in shortly

**Comment by gcallaghan-nd** [comment]
@ckifer , I see that my PR has been merged, thanks for that. Is there an ETA on the next package release?

**Comment by zendesk-sergey-germogentov** [comment]
@ckifer hi, any status updates on this please?

**Comment by ckifer** [comment]
@zendesk-sergey-germogentov anything specific you'd like an update on?

I believe we removed all the `img` roles so this shouldn't be an issue anymore. The contents of this thread have diverged a bit from the original issue

**Comment by zendesk-sergey-germogentov** [comment]
And this is the warning given by axe-core:

```
Selector: path[name="Critical"]
Rule ID: svg-img-alt
Issue: <svg> elements with an img role must have an alternative text
Description: Ensure <svg> elements with an img, graphics-document or graphics-symbol role have an accessible text
Recommendation: Fix any of the following: Element has no child that is a title aria-label attribute does not exist or is empty aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty Element has no title attribute
Impact: serious
Tags: cat.text-alternatives,wcag2a,wcag111,section508,section508.22.a,TTv5,TT7.a,EN-301-549,EN-9.1.1.1,ACT
HTML: <path cx="77.5" cy="75" name="Critical" stroke="#fff" fill="#CC3340" tabindex="-1" class="recharts-sector" d="M 132.5,75 A 55,55,0, 0,0, 132.47313016802562,73.28100042777466 L 107.48534372801397,74.06236386969528 A 30,30,0, 0,1, 107.5,75 Z" role="img">
```

**Comment by ckifer** [comment]
@zendesk-sergey-germogentov the `img` role is removed on Sectors in recharts@alpha (3.x)

There may be folks relying on these roles in tests, etc. so I think I originally treated this as a breaking change and still believe that it is. We're planning on releasing 3.0 soon

**Comment by ckifer** [comment]
In 2.x in order to get around the violation you can do something like this inside your `Pie`

```
  {data.map((entry, index) => (
          <Cell
            key={`cell-${index}`}
            aria-label="hello world"
          />
        ))}
```

**Comment by anithaKalaivanan** [comment]
Is this issue fixed for Scatter? If yes, which version would that be, as I still see it in 2.10.

**Comment by ckifer** [comment]
This will be fixed in 3.0 and should be fixed in the beta. It won't be fixed in 2.x

**Comment by pauldmps** [comment]
I am using v3.5.1 and the issue is still there with Pie charts.

<img width="1673" height="713" alt="Image" src="https://github.com/user-attachments/assets/7beb5754-b68e-4e55-a70e-d85b6b01368e" />

<img width="1673" height="713" alt="Image" src="https://github.com/user-attachments/assets/67d6deed-8e9f-4b4e-8f8c-f850f5e6cb9f" />
